web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java

```diff
@@ -0,0 +1,120 @@
+    private final SearchApiClient client;
+    private final Map<String, Object> optionalParameters;
+
+    /**
+     * @param apiKey             the Search API key for accessing their API
+     * @param timeout            the timeout duration for API requests
+     *                           <p>
+     *                           Default value is 30 seconds.
+     * @param engine             the engine used by Search API to execute the search
+     *                           <p>
+     *                           Default engine is Google Search.
+     * @param optionalParameters optional parameters to be passed on every request of this the engine, they can be overridden by the WebSearchRequest additional parameters for matching keys
+     *                           <p>
+     *                           Check <a href="https://www.searchapi.io">Search API</a> for more information on available parameters for each engine
+     */
+    @Builder
+    public SearchApiWebSearchEngine(String apiKey,
+                                    Duration timeout,
+                                    String engine,
+                                    Map<String, Object> optionalParameters) {
+        this.apiKey = ensureNotBlank(apiKey, "apiKey");
+        this.engine = getOrDefault(engine, DEFAULT_ENGINE);
+        this.optionalParameters = getOrDefault(optionalParameters, new HashMap<>());
+        this.client = SearchApiClient.builder()
+                .baseUrl(BASE_URL)
+                .timeout(getOrDefault(timeout, ofSeconds(30)))
+                .build();
+    }
+
+    @Override
+    public WebSearchResults search(WebSearchRequest webSearchRequest) {
+        SearchApiWebSearchRequest request = SearchApiWebSearchRequest.builder()
+                .apiKey(apiKey)
+                .engine(engine)
+                .query(webSearchRequest.searchTerms())
+                .optionalParameters(optionalParameters)
+                .additionalRequestParameters(webSearchRequest.additionalParams())
+                .build();
+        SearchApiWebSearchResponse response = client.search(request);
+        return toWebSearchResults(response);
+    }
+
+    private WebSearchResults toWebSearchResults(SearchApiWebSearchResponse response) {
+        List<OrganicResult> organicResults = response.getOrganicResults();
+        Long totalResults = ((Integer) organicResults.size()).longValue(); // not ideal, but it may not be present in the response and is required not null by WebSearchInformationResult
+        WebSearchInformationResult searchInformation = WebSearchInformationResult.from(
+                totalResults,
+                getCurrentPage(response.getPagination()),
+                null
+        );
+        Map<String, Object> searchMetadata = getOrDefault(response.getSearchParameters(), new HashMap<>());
+        addToMetadata(searchMetadata, response.getPagination());
+        addToMetadata(searchMetadata, response.getSearchInformation());
```

Comment:
I think that `.getSearchInformation()` have own field and `.getPagination()` not is really relevant.
```suggestion
        Map<String, Object> searchMetadata = getOrDefault(response.getSearchMetadata(), new HashMap<>());
        addToMetadata(searchMetadata, response.getSearchParameters());
```

Reply:
@czelabueno it was requested by @mark-searchapi to add everything to the metadata, again because every engine may have different namings.

Reply:
Yes, I understand it, but let's not lose sight of the fact that it will be used as a `ContentRetriever` or `Tool` by an LLM, so it is necessary to analyze if it is relevant information to index it to the LLM in order to provide better responses to the user.
This [comment](https://github.com/langchain4j/langchain4j/pull/642/files#r1566496331) explains the current use of the metadata  and for now we are only extracting content from the `WebSearchOrganicResult`. In version 2, we will add the `WebSearchInformationResult` data so any other relevant content to be ingested to the LLM can be included there. `WebSearchResults.searchMetadata` would be used for debugging purpose.
The pagination is not considered content that adds any value to LLM for now and it could be included here `WebSearchResults.searchMetadata`

Reply:
I wasn't aware of metadata use cases. Since `WebSearchResults.searchMetadata` is used for debugging, we can put everything that is in our `search_metadata` root key. It includes information about search execution time and has links to the JSON/HTML for debugging purposes.

Reply:
@mark-searchapi In yours Baidu docs it doesn't have `search_metadata`, is it optional?